src/main.cpp

```diff
@@ -1,19 +1,234 @@
+        this->create_subscription<interfaces::msg::MissionProgress>(
+            common_lib::topic_names::MissionProgress, 10,
+            std::bind(&WaypointNode::callback_mission_progress, this, _1));
+
+    // Create a publisher for the "uav_waypoint_command" topic
+    uav_waypoint_command_publisher =
+        this->create_publisher<interfaces::msg::UAVWaypointCommand>(
+            common_lib::topic_names::UAVWaypointCommand, 10);
+
+    // Initialize Event Loop
+    event_loop_timer = this->create_wall_timer(
+        std::chrono::milliseconds(event_loop_time_delta_ms),
+        std::bind(&WaypointNode::event_loop, this));
+}
+
+/**
+ * @brief Checks if the first loop was already triggered and returns the result.
+ *
+ * This function checks the state of the first loop and returns false if it was
+ * already triggered, otherwise it returns true. After returning the state, it
+ * sets the first loop state to false.
+ *
+ * @return False if the first loop was already triggered, true otherwise.
+ */
+bool WaypointNode::get_state_first_loop() {
+    // Check if first loop was already triggered
+    if (!state_first_loop) return false;
+
+    // Set first loop to false and return true
+    state_first_loop = false;
+    return true;
+}
+
+/**
+ * Checks if the current mission is finished.
+ *
+ * @return true if the mission is finished, false otherwise.
+ */
+bool WaypointNode::current_mission_finished() {
+    if (mission_progress >= 1.0) {
+        mission_progress = 0.0;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/**
+ * @brief Executes the event loop for the WaypointNode.
+ *
+ * This function is responsible for executing the event loop of the WaypointNode
+ * class. It checks the node state and performs the corresponding actions based
+ * on the state. The possible states are:
+ * - init: Initializes the mode.
+ * - pre_wait_time: Wait a predefined time.
+ * - reach_cruise_height: Fly drone to the cruising height.
+ * - fly_to_waypoint: Executes the mode to fly to the waypoint.
+ * - reach_target_height: Fly drone to the target height.
+ * - post_wait_time: Wait a predefined time.
+ *
+ * If the node state is unknown, an error message is logged and a job_finished
+ * error message is sent.
+ */
+void WaypointNode::event_loop() {
+    if (!this->get_active()) return;
+
+    switch (get_node_state()) {
+        case init:
+            mode_init();
+            break;
+        case pre_wait_time:
+            break;  // Wait for the `wait_timer` to trigger
+        case reach_cruise_height:
+            mode_reach_cruise_height();
+            break;
+        case fly_to_waypoint:
+            mode_fly_to_waypoint();
+            break;
+        case reach_target_height:
+            mode_reach_target_height();
+            break;
+        case post_wait_time:
+            break;  // Wait for the `wait_timer` to trigger
+        default:
+            RCLCPP_ERROR(this->get_logger(),
```

Comment:
I would consider this a grave error and exit the process. But that´s up to you

Reply:
Agree